### DISCUSSION TITLE
Best practice for pulling to refresh an entire screen

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2204] [Q&A] [answered]
Long-story short: how can I refetch a query from its parent?

Imagine I want to pull to refresh all the queries on a given screen. What would be the best way to do that? I'm thinking specifically for a React Native app. I'm fetching a lot of data inside of nested components rather than at the root screen level. 

Should I just copy all the hooks from nested children up to the top-level screen view, and combine their `result.fetching` variables there? The only other thing I could think of is to "register" from the children components, but that feels like a really risky and overall bad practice that goes against react principles.

This question probably isn't super urql-specific, but maybe there's some sort of graphql-y answer to it.

Another solution is to combine all the screen's queries into one query, put it at the root of the screen with `pause: true`, and then call that when someone pulls to refresh, triggering all the items below it to re-render on an update. This would also be beneficial since re-renders would only occur at the root and not at the children while `fetching` is `true`.

However, for child components that conditionally fetch data based on some state, this likely isn't feasible.

My guess is that the advice will be to lift up state with a reusable hook for the entire screen, but I figured I'd see if there is any advice for this pattern.

Thanks!

**Comment by erquhart** [comment]
I landed on a solution for this that I'm pretty satisfied with - use a context and register a refresh function and a boolean to represent the refreshing state for each query. If you have any kind of centralized hook that you run with your useQuery calls (or if you wrap useQuery with your own hook), you can implement in one place. Even better if you have a shared container with your pull to refresh logic. Doesn't seem to be a lot of folks looking at this discussion but if anyone wants more details I'm happy to help, just @ me.

Update: shouldn't wrap your app as all mounted screens will refresh every time - wrap the context around the current screen (again, a shared screen container helps).

**Reply by nandorojo** [reply]
is there a way to wrap functions from graphql codegen somehow? that’s what i’m using. this approach makes sense, it’s interesting to register it upwards from the hooks themselves

**Reply by erquhart** [reply]
I'm sure it's possible, you'd need to create the hook and context in your app and then maybe write a plugin that adds the hook to each query. I have not used graphql codegen at all, there may be an even simpler way than writing a plugin.

**Comment by tgriesser** [comment]
Not react specific, but you could use a custom exchange, similar to the [refocusExchange](https://github.com/FormidableLabs/urql/blob/main/exchanges/refocus/src/refocusExchange.ts), which registers the queries, and then expose a "refresh" function you can call to re-execute them:


```ts
import { pipe, tap } from 'wonka';
import { Exchange, Operation, Client } from '@urql/core';

export const makeRefreshQueriesExchange = (): Exchange => {
  const watchedOperations = new Map<number, Operation>();
  const observedOperations = new Map<number, number>();

  let urqlClient: Client

  // Refreshes all of the mounted queries
  function refresh() {
    if (urqlClient) {
      watchedOperations.forEach(op => {
        urqlClient.reexecuteOperation(
          urqlClient.createRequestOperation('query', op, {
            ...op.context,
            requestPolicy: 'cache-and-network',
          })
        );
      });
    }
  }

  const refreshExchange = ({ client, forward }) => ops$ => {
    urqlClient = client
    const processIncomingOperation = (op: Operation) => {
      if (op.kind === 'query' && !observedOperations.has(op.key)) {
        observedOperations.set(op.key, 1);
        watchedOperations.set(op.key, op);
      }

      if (op.kind === 'teardown' && observedOperations.has(op.key)) {
        observedOperations.delete(op.key);
        watchedOperations.delete(op.key);
      }
    };
    return forward(pipe(ops$, tap(processIncomingOperation)));
  };
  return { refreshExchange, refresh }
};
```

Then to use:

```ts
const { refreshExchange, refresh } = makeRefreshQueriesExchange()

export { refresh }

urqlClient({
  exchanges: [
    dedupExchange,
    refreshExchange,
    cacheExchange,
    fetchExchange
  ]
})
```

Then, when you need to refresh the queries, call:

```
refresh()
```

**Reply by nandorojo** [reply]
Thanks for writing this! Is there a way you'd recommend tracking the `refreshing` state? This would be a more react-y question. perhaps it just has to update React state with a listener?

The issue on my end is I have no idea how these exchanges work (or even more, how wonka works with its peculiar syntax and JS approach)

**Reply by tgriesser** [reply]
> This would be a more react-y question. perhaps it just has to update React state with a listener?

The `stale` prop of the operation result returned from `useQuery` should be set to `true` while it's doing the `cache-and-network` call: https://github.com/FormidableLabs/urql/blob/e1278c91b8bb601e57835179c202ac4526c6f1bf/exchanges/graphcache/src/cacheExchange.ts#L323-L328 not sure, it might also set the `fetching` prop for you, that the react hooks appear to provide.

> The issue on my end is I have no idea how these exchanges work (or even more, how wonka works with its peculiar syntax and JS approach)

Yeah, it's a little bit of a learning curve, but since the React integration works layered on top of the exchanges, it should "just work"

https://formidable.com/open-source/urql/docs/architecture/#the-client-and-exchanges

**Reply by erquhart** [reply]
I'm using `fetching || stale` for refreshing state, been working really well. They're guaranteed to both settle to false eventually.

**Reply by erquhart** [reply]
Another challenge to consider with the exchange approach is that refreshing will refresh all queries for mounted components, so if you've navigated a few screens they all get refreshed. This is one of the few cases where React contexts felt like an elegant solution.

**Reply by nandorojo** [reply]
yeah that’s true, especially for react native where hidden screens can remain mounted (unless you’re using react-freeze I believe)

**Reply by erquhart** [reply]
Ah, didn't know about react-freeze, good tip!

**Reply by tgriesser** [reply]
Yeah I assumed refresh all queries on the screen meant all mounted queries. If that’s not the case you’d probably want to use a different approach

**Reply by kitten** [reply]
If this is only a concern of refreshing from a parent to a child component btw, you could consider providing some sort of signal:

```js
const [signal, trigger] = useState(() => {
  const fns = []
  return [
    function signal(callback) {
      fns.push(callback)
      return () => {
        const index = fns.indexOf(callback)
        if (index > -1) fns.splice(index, 1)
      }
    },
    function trigger() {
      fns.slice().forEach(fn => fn())
    },
  ]
})
```

I haven't tested this but if it's just for a repeating pattern it'd then be pretty straightforward to use:

```js
useEffect(() => {
  return signal(() => {
    reexecuteQuery({ requestPolicy: 'network-only' }))
  })
}, [signal, reexecuteQuery])
```

just an idea though ✌️

**Reply by kitten** [reply]
To be fair, I'd still say though, if multiple queries on the screen only exist to manage separate data dependencies, composing fragments and having a single query is still best in my opinion; at least, that's the Relay way of doing things

**Reply by nandorojo** [reply]
yeah, I just want to avoid prop drilling. are you suggesting using the same fragment at the top level of a screen and also in sub components with shared fragments, or just doing one query at the top of a screen and passing down the result?

I’m refactoring everything to graphql so I want to get a handle on the right patterns.

**Comment by nandorojo** [comment, marked as answer]
To anyone still curious about this: even when I put all my data requirements into one query, I often get flickering between `fetching` and `stale` becoming `true`/`false`. I ultimately decided to "fake" the pull to refresh by adding it as a custom function in my `useQuery` wrapper. It shows a refreshing indicator for one second, and then goes away. It is too jarring otherwise on React Native.

```tsx
const [{ fetching, stale, data, error, operation }, executeQuery] = useQuery({
    // @ts-expect-error it's fine
    query: document,
    ...rest,
  })

  const client = useClient()

  const [isPullingToRefresh, setPullingToRefresh] = useState(false)

  const timer = useRef(0)

  const onPullToRefresh = useLatestCallback(() => {
    clearTimeout(timer.current)
    setPullingToRefresh(true)
    executeQuery({
      requestPolicy: 'network-only',
    })
    new Promise<void>((resolve) => {
      timer.current = setTimeout(() => {
        clearTimeout(timer.current)
        resolve()
      }, 1000)
    })
      .catch()
      .finally(() => {
        setPullingToRefresh(false)
      })
  })

  useEffect(
    () => () => {
      clearTimeout(timer.current)
    },
    []
  )

  return {
    fetching,
    stale,
    data,
    error,
    variables: operation?.variables,
    isPullingToRefresh,
    onPullToRefresh,
    execute: useCallback(async () => {
      return executeQuery({
        requestPolicy: 'network-only',
      })
    }, [executeQuery]),
  }
```
